src/analysis/special_functions/bernstein.lean

```diff
@@ -0,0 +1,408 @@
+for a continuous function `f : C([0,1], ℝ)` converge uniformly to `f` as `n` tends to infinity.
+
+Our proof follows [Richard Beals' *Analysis, an introduction*][beals-analysis], §7D.
+The original proof, due to [Bernstein](bernstein1912) in 1912, is probabilistic,
+and relies on Bernoulli's theorem,
+which gives bounds for how quickly the observed frequencies in a
+Bernoulli trial approach the underlying probability.
+
+The proof here does not directly rely on Bernoulli's theorem,
+but can also be given a probabilistic account.
+* Consider a weighted coin which with probability `x` produces heads,
+  and with probability `1-x` produces tails.
+* The value of `bernstein n k x` is the probability that
+  such a coin gives exactly `k` heads in a sequence of `n` tosses.
+* If such an appearance of `k` heads results in a payoff of `f(k / n)`,
+  the `n`-th Bernstein approximation for `f` evaluated at `x` is the expected payoff.
+* The main estimate in the proof bounds the probability that
+  the observed frequency of heads differs from `x` by more than some `δ`,
+  obtaining a bound of `(4 * n * δ^2)⁻¹`, irrespective of `x`.
+* This ensures that for `n` large, the Bernstein approximation is (uniformly) close to the
+  payoff function `f`.
+
+(You don't need to think in these terms to follow the proof below: it's a giant `calc` block!)
+
+This result proves Weierstrass' theorem that polynomials are dense in `C([0,1], ℝ)`,
+although we defer an abstract statement of this until later.
+-/
+
+
+noncomputable theory
+
+open_locale classical
+open_locale big_operators
+open_locale bounded_continuous_function
+
+lemma division_ring.mul_left_injective {α : Type*} [division_ring α] {x : α} (h : x ≠ 0) :
+  function.injective (λ y, x * y) :=
+λ y y' w, begin
+  apply_fun (λ y, x⁻¹ * y) at w,
+  simpa only [←mul_assoc, inv_mul_cancel h, one_mul] using w,
+end
+
+lemma division_ring.mul_right_injective {α : Type*} [division_ring α] {x : α} (h : x ≠ 0) :
+  function.injective (λ y, y * x) :=
+λ y y' w, begin
+  apply_fun (λ y, y * x⁻¹) at w,
+  simpa only [mul_assoc, mul_inv_cancel h, mul_one] using w,
+end
+
+lemma mul_unit_interval_le {α : Type*} [ordered_semiring α] {a b c : α}
+  (h₁ : 0 ≤ c) (h₂ : a ≤ c) (h₃ : 0 ≤ b) (h₄ : b ≤ 1) : a * b ≤ c :=
+begin
+  conv_rhs { rw ←mul_one c, },
+  exact mul_le_mul h₂ h₄ h₃ h₁,
+end
+
+lemma finset.sum_le_univ_sum_of_nonneg
+  {α β : Type*} [fintype α] {s : finset α} [ordered_cancel_add_comm_monoid β] {f : α → β}
+  (w : ∀ a, 0 ≤ f a) :
+∑ a in s, f a ≤ ∑ a, f a :=
+begin
+  rw ←finset.sum_add_sum_compl s,
+  exact le_add_of_nonneg_right (finset.sum_nonneg (λ x m, w x)),
+end
+
+lemma pow_minus_two_nonneg {α : Type*} [linear_ordered_field α] (a : α) : 0 ≤ a^(-2 : ℤ) :=
+begin
+  simp only [inv_nonneg, fpow_neg],
+  apply pow_two_nonneg,
+end
+
+
+/--
+The Bernstein polynomials, as continuous functions on ℝ.
+-/
+def bernstein' (n ν : ℕ) : C(ℝ, ℝ) :=
+⟨λ x, polynomial.aeval x (bernstein_polynomial n ν), by continuity⟩
```

Comment:
I agree! I think it might even be best to include redesigning `X →ᵇ Y` so it is a structure extending `C(X, Y)` as part of this project.

I'd propose considering that refactor as orthogonal to this PR --- I would like to make progress on it, because I think it will be helpful for stating the abstract version of the Weierstrass theorem, and for stating and proving Stone-Weierstrass.

This PR will probably take a while still before it's ready to merge, as there are several dependent PRs. If this proposed refactor makes it in time, then I'll rewrite `bernstein'` as you describe. If not, it can happen later.

Reply:
Sounds good!

Reply:
And feel free to assign me part of that refactor.